### PR TITLE
Updated the driver to be based on the Reactive Streams Java driver

### PR DESCRIPTION
**Pull request by rozza**

### CODE REVIEW

**Comment by rozza** [comment]
@jyemin it's probably easier to ignore the diff and just look at the code + tests.

File: driver/src/main/scala/com/mongodb/scala/reactivestreams/client/MongoClient.scala [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+import com.mongodb.ConnectionString
+import com.mongodb.async.client.MongoClientOptions
+import com.mongodb.reactivestreams.client.{ MongoClient => JMongoClient, MongoClients }
+import org.bson.Document
+import org.reactivestreams.Publisher
+
+import scala.reflect.ClassTag
+
+object MongoClient {
```
Comment (jyemin) [inline driver/src/main/scala/com/mongodb/scala/reactivestreams/client/MongoClient.scala:29]: Does this need a doc comment?

File: driver/src/main/scala/com/mongodb/scala/reactivestreams/client/MongoClient.scala [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+  def listDatabaseNames(): Publisher[String] = wrapped.listDatabaseNames()
+
+  /**
+   * Gets the list of databases
+   *
+   * @param clazz the class to cast the database documents to
+   * @tparam T   the type of the class to use instead of `Document`.
+   * @return the fluent list databases interface
+   */
+  def listDatabases[T]()(implicit e: T DefaultsTo Document, ct: ClassTag[T]): ListDatabasesPublisher[T] =
```
Comment (jyemin) [inline driver/src/main/scala/com/mongodb/scala/reactivestreams/client/MongoClient.scala:104]: Are you committing to having org.bson.Document be the default T, or is this a placeholder?

File: driver/src/main/scala/com/mongodb/scala/reactivestreams/client/MongoClient.scala [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+   *
+   * [[http://docs.mongodb.org/manual/reference/commands/listDatabases List Databases]]
+   * @return an iterable containing all the names of all the databases
+   */
+  def listDatabaseNames(): Publisher[String] = wrapped.listDatabaseNames()
+
+  /**
+   * Gets the list of databases
+   *
+   * @param clazz the class to cast the database documents to
```
Comment (jyemin) [inline driver/src/main/scala/com/mongodb/scala/reactivestreams/client/MongoClient.scala:100]: is 'clazz' a thing in the API?  If not, remove from the docs.

File: driver/src/main/scala/com/mongodb/scala/reactivestreams/client/MongoDatabase.scala [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+   */
+  def executeCommand[T](command: AnyRef)(implicit e: T DefaultsTo Document, ct: ClassTag[T]): Publisher[T] =
+    wrapped.executeCommand[T](command, ct.runtimeClass.asInstanceOf[Class[T]])
+
+  /**
+   * Executes command in the context of the current database.
+   *
+   * @param command        the command to be run
+   * @param readPreference the { @link com.mongodb.ReadPreference} to be used when executing the command
+   * @param clazz          the default class to cast any documents returned from the database into.
```
Comment (jyemin) [inline driver/src/main/scala/com/mongodb/scala/reactivestreams/client/MongoDatabase.scala:115]: errant clazz?

File: driver/src/main/scala/com/mongodb/scala/reactivestreams/client/MongoDatabase.scala [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+   * @tparam T            the type of the class to use instead of { @code Document}.
```
Comment (jyemin) [inline driver/src/main/scala/com/mongodb/scala/reactivestreams/client/MongoDatabase.scala:116]: should the description line up?

File: driver/src/main/scala/com/mongodb/scala/reactivestreams/client/MongoDatabase.scala [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+  /**
+   * Create a new collection with the selected options
+   *
+   * [[http://docs.mongodb.org/manual/reference/commands/create Create Command]]
+   * @param collectionName the name for the new collection to create
+   * @param options        various options for creating the collection
+   * @return a publisher identifying when the collection has been created
+   */
+  def createCollection(collectionName: String, options: CreateCollectionOptions): Publisher[Void] =
+    wrapped.createCollection(collectionName, options)
```
Comment (jyemin) [inline driver/src/main/scala/com/mongodb/scala/reactivestreams/client/MongoDatabase.scala:165]: The dreaded Publisher[Void]

File: driver/src/main/scala/com/mongodb/scala/reactivestreams/client/MongoCollection.scala [hunk 7]
```diff
@@ -0,0 +1,479 @@
...
+   * @return a publisher emitting the sequence of distinct values
+   */
+  def distinct[C](fieldName: String)(implicit ct: ClassTag[C]): DistinctPublisher[C] =
+    DistinctPublisher(wrapped.distinct(fieldName, ct))
+
+  /**
+   * Finds all documents in the collection.
+   *
+   * [[http://docs.mongodb.org/manual/tutorial/query-documents/ Find]]
+   * @param clazz the class to decode each document into
```
Comment (jyemin) [inline driver/src/main/scala/com/mongodb/scala/reactivestreams/client/MongoCollection.scala:148]: errant clazz?  I'll stop pointing these out, as they may not be errant, and there's a lot of them.

File: driver/src/main/scala/com/mongodb/scala/reactivestreams/client/MongoClient.scala (hunk 1 above)
Comment (rozza) [inline driver/src/main/scala/com/mongodb/scala/reactivestreams/client/MongoClient.scala:29]: Its a companion object and the api the docs default to the case class - but have added some small doc text.

File: driver/src/main/scala/com/mongodb/scala/reactivestreams/client/MongoClient.scala (hunk 2 above)
Comment (rozza) [inline driver/src/main/scala/com/mongodb/scala/reactivestreams/client/MongoClient.scala:104]: No its just the default at the moment - I'm spiking a custom document and BsonDocument implementations.

File: driver/src/main/scala/com/mongodb/scala/reactivestreams/client/MongoClient.scala (hunk 3 above)
Comment (rozza) [inline driver/src/main/scala/com/mongodb/scala/reactivestreams/client/MongoClient.scala:100]: weirdly the scaladoc doesn't complain but just strips them and enabling the check in scalastyle shows a bug as scalastyle is providing false positives.  Have removed clazz and will reenable scaladoc checks in scalastyle once the bugs fixed.

File: driver/src/main/scala/com/mongodb/scala/reactivestreams/client/MongoDatabase.scala (hunk 4 above)
Comment (rozza) [inline driver/src/main/scala/com/mongodb/scala/reactivestreams/client/MongoDatabase.scala:115]: Done

File: driver/src/main/scala/com/mongodb/scala/reactivestreams/client/MongoDatabase.scala (hunk 5 above)
Comment (rozza) [inline driver/src/main/scala/com/mongodb/scala/reactivestreams/client/MongoDatabase.scala:116]: Dont follow - but I removed clazz

File: driver/src/main/scala/com/mongodb/scala/reactivestreams/client/MongoDatabase.scala (hunk 6 above)
Comment (rozza) [inline driver/src/main/scala/com/mongodb/scala/reactivestreams/client/MongoDatabase.scala:165]: confirmed to be not allowed in RC2

File: driver/src/main/scala/com/mongodb/scala/reactivestreams/client/MongoCollection.scala (hunk 7 above)
Comment (rozza) [inline driver/src/main/scala/com/mongodb/scala/reactivestreams/client/MongoCollection.scala:148]: yeap - removed

**Comment by rozza** [comment]
Fixed rebased and merged

**Pull request closed without merging**